File: swift/06_ZigzagConversation.swift
class Solution {
    func convert(_ s: String, _ numRows: Int) -> String {
        if numRows == 1 {
            return s
        }
        
        if s.count <= 1 {
            return s
        }
        
        var rows = [String] (repeating: String(), count: min(s.count, numRows))
        var curRow = 0
        var isDown = false
        
        for ch in s {
            rows[curRow] += String(ch)
            if curRow == 0 || curRow == numRows - 1  {
                isDown = !isDown
            }
            curRow += isDown ? 1 : -1
        }
        
        return rows.joined()
    }
}

File: java/06_ZigzagConversation.java
// https://leetcode.com/problems/zigzag-conversion/
class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1) {
            return s;
        }

        List<StringBuilder> rows = new ArrayList<>();

        for (int i = 0; i < Math.min(numRows, s.length()); i++) {
            rows.add(new StringBuilder());
        }

        int curRow = 0;
        boolean isDown = false;

        for (char c : s.toCharArray()) {
            rows.get(curRow).append(c);
            if (curRow == 0 || curRow == numRows - 1) {
                isDown = !isDown;
            }
            curRow += isDown ? 1 : - 1;
        }

        StringBuilder result = new StringBuilder();
        for (StringBuilder row : rows) {
            result.append(row);
        }

        return result.toString();
    }
}